src/test/java/pl/widulinski/webcrawlertool/ScrAppApplicationTests.java

```diff
@@ -0,0 +1,205 @@
+        }
+
+        assertEquals("Allegro - atrakcyjne ceny i 100% bezpieczeństwa", driver.getTitle());  //testing does opened page using webdriver
+
+
+        driver.close();
+    }
+
+
+    @Test
+    @Order(4)
+    public void getDataFromRepositoryTest() {
+
+
+        for (DataToScrap element : dataToScrapRepository.findAll()) {
+
+            elements.put(element.getShop() + "_" + element.getCategory(), element);
+        }
+        assertFalse(elements.isEmpty());
+    }
+
+
+    @Test
+    @Order(5)
+    public void getLastPageTest() {
+
+        System.setProperty("webdriver.chrome.driver", "C:\\chromedriver.exe");
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("headless");
+        WebDriver driver = new ChromeDriver(options);
+
+        elements.forEach((key, value) -> {
+
+            driver.get(value.getUrlToCategory() + "1");
+
+            driver.manage().timeouts().implicitlyWait(3, TimeUnit.SECONDS);
+
+            WebElement webElement = driver.findElement(By.xpath(value.getXpathOfLastPage().replace("|", "'")));
+
+            assertTrue(webElement.isEnabled());
+
+        });
+
+        driver.close();
+    }
+
+    @Test
+    @Order(6)
+    public void scrapArticleDataTest() {
+
+        System.setProperty("webdriver.chrome.driver", "C:\\chromedriver.exe");
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("headless");
+        WebDriver driver = new ChromeDriver(options);
+
+        DataToScrap foundWebElement = dataToScrapRepository.findByShopAndCategory("Allegro", Categories.ELECTRONICS);
+
+        driver.get(foundWebElement.getUrlToCategory() + "1");
+
+
+
+        //replacing signs '|' to '''
+        String replacedArticleElement = foundWebElement.getXpathToArticleElement().replace("|", "'");
+        String replacedArticleName = foundWebElement.getXpathToArticleName().replace("|", "'");
+        String replacedArticleHref = foundWebElement.getXpathToArticleHref().replace("|", "'");
+        String replacedArticlePrice = foundWebElement.getXpathToArticlePrice().replace("|", "'");
+        //end of replacing
+
+        List<WebElement> webElementsList = driver.findElements(By.xpath(replacedArticleElement));
+
+
+        for (WebElement element : webElementsList
+        ) {
+
+            webElementTest.setName(element.findElement(By.xpath("." + replacedArticleName)).getText());
+            webElementTest.setLink(element.findElement(By.xpath("." + replacedArticleHref)).getAttribute("href"));
+            webElementTest.setPrice(element.findElement(By.xpath("." + replacedArticlePrice)).getText());
+
+            assertFalse(webElementTest.getName().isEmpty());
+            assertFalse(webElementTest.getLink().isEmpty());
+            assertFalse(webElementTest.getPrice().isEmpty());
+        }
+
+        driver.close();
+
+    }
+
+    @Test
+    @Order(7)
+    public void createExcelTest() throws IOException {
```

Comment:
Tutaj testujesz zapisywanie do Excela. Klasa ktora to robi to `CreateExcel` wiec utworz osobna klase testowa `CreateExcelTest` i przetestuj tak jak tutaj z tym że stream powinienes wypełnic a potem zerknac do pliku czy to co w srodku jest Cie zadawala. To moze być ciezkie jesli jest w formacie xls a nie csv - ale moim zdanie warto by wtedy uzyc apache POI zeby sprawdzic czy wartosci w srodku sa odpowiednie. Grzesiek W. mial podobny case - zobacz jak on testuje Excela. https://github.com/GriszaKaramazow/currency-cmdline/blob/master/src/test/java/pl/connectis/print/XLSXPrinterTests.java
https://github.com/GriszaKaramazow/currency-cmdline/blob/master/src/test/java/pl/connectis/utils/ExcelFileTestHelper.java
Dorobil pare metod ktore wyciagaja dane z konkretnych komórek i sprawdza wartosci

Reply:
poprawiłem. Mam nadzieję, że chociaż po części będzie OK.